core/src/bls_sigverify/sigverifier.rs

```diff
@@ -0,0 +1,1681 @@
+        let verified_certs = certs_to_verify
+            .into_par_iter()
+            .filter(
+                |cert_to_verify| match self.verify_bls_certificate(cert_to_verify, bank) {
+                    Ok(()) => true,
+                    Err(e) => {
+                        trace!(
+                            "Failed to verify BLS certificate: {:?}, error: {e}",
+                            cert_to_verify.cert_type
+                        );
+                        self.stats
+                            .received_bad_signature_certs
+                            .fetch_add(1, Ordering::Relaxed);
+                        false
+                    }
+                },
+            )
+            .collect();
+        certs_batch_verify_time.stop();
+        self.stats
+            .certs_batch_elapsed_us
+            .fetch_add(certs_batch_verify_time.as_us(), Ordering::Relaxed);
+        verified_certs
+    }
+
+    fn verify_bls_certificate(
+        &self,
+        cert_to_verify: &Certificate,
+        bank: &Bank,
+    ) -> Result<(), CertVerifyError> {
+        if self
+            .verified_certs
+            .read()
+            .unwrap()
+            .contains(&cert_to_verify.cert_type)
+        {
+            self.stats.received_verified.fetch_add(1, Ordering::Relaxed);
+            return Ok(());
+        }
+
+        let slot = cert_to_verify.cert_type.slot();
+        let Some(key_to_rank_map) = get_key_to_rank_map(bank, slot) else {
+            return Err(CertVerifyError::KeyToRankMapNotFound(slot));
+        };
+
+        let max_len = key_to_rank_map.len();
+
+        let decoded_bitmap = match decode(&cert_to_verify.bitmap, max_len) {
+            Ok(decoded) => decoded,
+            Err(e) => {
+                return Err(CertVerifyError::BitmapDecodingFailed(e));
+            }
+        };
+
+        match decoded_bitmap {
+            solana_signer_store::Decoded::Base2(bit_vec) => {
+                self.verify_base2_certificate(cert_to_verify, &bit_vec, key_to_rank_map)?
+            }
+            solana_signer_store::Decoded::Base3(bit_vec1, bit_vec2) => self
+                .verify_base3_certificate(cert_to_verify, &bit_vec1, &bit_vec2, key_to_rank_map)?,
+        }
+
+        self.verified_certs
+            .write()
+            .unwrap()
+            .insert(cert_to_verify.cert_type);
+
+        Ok(())
+    }
+
+    fn verify_base2_certificate(
+        &self,
+        cert_to_verify: &Certificate,
+        bit_vec: &BitVec<u8, Lsb0>,
+        key_to_rank_map: &Arc<BLSPubkeyToRankMap>,
+    ) -> Result<(), CertVerifyError> {
+        let original_vote = cert_to_verify.cert_type.to_source_vote(true).ok_or(
+            CertVerifyError::NoSourceVoteFound(cert_to_verify.cert_type, true),
+        )?;
+
+        let Ok(signed_payload) = bincode::serialize(&original_vote) else {
+            return Err(CertVerifyError::SerializationFailed);
+        };
```

Comment:
It seems like [sometimes](https://github.com/anza-xyz/agave/pull/8929/files#diff-82d08cb2b87b1afa1b1fbe8e060de71d1939794cf19bc35808ba9e6e0fc844b2R683) we panic if this serialization fails and sometimes we don't?